common/persistence/cassandra/cassandraHistoryPersistence.go

```diff
@@ -80,13 +144,989 @@ func newHistoryPersistence(cfg config.Cassandra, logger bark.Logger) (p.HistoryS
+			break GetFailureReasonLoop
+		}
+
+		treeID := previous["tree_id"].(gocql.UUID).String()
+		branchID := previous["branch_id"].(gocql.UUID).String()
+		nodeID := previous["node_id"].(int64)
+
+		if rowType == rowTypeHistoryNode && treeID == reqTreeID && branchID == rootNodeBranchID && nodeID == rootNodeID {
+			actualTxnID = previous["txn_id"].(int64)
+			if actualTxnID != reqTxnID {
+				txnIDUnmatch = true
+			}
+		} else if rowType == rowTypeHistoryBranch && treeID == reqTreeID && branchID == reqBranchID {
+			branchAlreadyExits = true
+		}
+
+		previous = make(map[string]interface{})
+		if !iter.MapScan(previous) {
+			break GetFailureReasonLoop
+		}
+	}
+
+	if txnIDUnmatch || branchAlreadyExits {
+		return &p.ConditionFailedError{
+			Msg: fmt.Sprintf("Failed to create a new branch. txnIDUnmatch: %v, branchAlreadyExits: %v . Request txn_id: %v, Actual Value: %v",
+				txnIDUnmatch, branchAlreadyExits, reqTxnID, actualTxnID),
+		}
+	}
+
+	// At this point we only know that the write was not applied.
+	var columns []string
+	columnID := 0
+	for _, previous := range allPrevious {
+		for k, v := range previous {
+			columns = append(columns, fmt.Sprintf("%v: %s=%v", columnID, k, v))
+		}
+		columnID++
+	}
+	return &p.UnexpectedConditionFailedError{
+		Msg: fmt.Sprintf("Failed to create a new branch. Request txn_id: %v, Actual Value: %v, columns: (%v)",
+			reqTxnID, actualTxnID, columns),
+	}
+}
+
+// return error is branch is already deleted/not exists
+func (h *cassandraHistoryPersistence) validateBranchStatus(treeID string, branchID string) error {
+	query := h.session.Query(v2templateReadOneNode,
+		treeID,
+		branchID,
+		rowTypeHistoryBranch,
+		branchNodeID)
+
+	result := make(map[string]interface{})
+	if err := query.MapScan(result); err != nil {
+		return convertCommonErrors("validateBranchStatus", err)
+	}
+
+	deleted := result["deleted"].(bool)
+	if deleted {
+		return &workflow.EntityNotExistsError{
+			Message: fmt.Sprintf("branch %v of tree %v is already deleted", branchID, treeID),
+		}
+	}
+	return nil
+}
+
+func (h *cassandraHistoryPersistence) createRoot(treeID string) (bool, error) {
+	var query *gocql.Query
+
+	query = h.session.Query(v2templateInsertNode,
+		treeID,
+		rootNodeBranchID,
+		rowTypeHistoryNode,
+		nil, // ancestors
+		false,
+		rootNodeID,
+		initialTransactionID,
+		nil, //data
+		nil) //data_encoding
+
+	previous := make(map[string]interface{})
+	applied, err := query.MapScanCAS(previous)
+	if err != nil {
+		return false, convertCommonErrors("createRoot", err)
+	}
+
+	return applied, nil
+}
+
+// AppendHistoryNodes add(or override) a batch of nodes to a history branch
+// Note that it's not allowed to override the ancestors' nodes, which means NextNodeIDToUpdate > forking point
+func (h *cassandraHistoryPersistence) AppendHistoryNodes(request *p.InternalAppendHistoryNodesRequest) (*p.InternalAppendHistoryNodesResponse, error) {
+	branchInfo, err := h.refillAncestors(request.BranchInfo)
```

Comment:
See if we can optimize this using a cache of ancestors on mutable state cache on history service.